### PR TITLE
[RFC][Tx Sort] Implement sorting of inputs

### PR DESCRIPTION
**Pull request by savil**
Follows BIP69: https://github.com/bitcoin/bips/blob/master/bip-0069.mediawiki

1. Implements sorting of transaction inputs.
- BIP says to use "reversed byte-order" for the `prev_hash`. I interpreted this as: little-endian.
- TODO need to add tests

2. Re: improve sorting of TxOut's script_pubkey to use lexicographic ordering, and not length.
From the test-cases i've included it seems that the current code already does lexicographic ordering (and not length based). Am i missing something?

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Cool, thanks. Can you add the test vectors from the BIP (and add some test vectors to the BIP that test variable-length scriptPubKeys)?

**Comment by savil** [comment]
@TheBlueMatt sure, yeah happy to, but I've been running into some issues.

For example, the given values in the BIP don't seem to serialize-deserialize cleanly. See the test output at the bottom of https://gist.github.com/savil/d6acec10aa5a8ea6698fbb19443d68cb.  Not yet sure if that indicates a bug in the hex-string or in the rust-bitcoin's serialize/deserialize implementation.

Also, looks like `rustc-serialize` is now deprecated, so looking into using the `hex` library, https://users.rust-lang.org/t/deprecation-of-rustc-serialize/10485 but that one cannot deserialize the other hex-string in the BIP cleanly. https://gist.github.com/savil/dbe9963a263baba4d32de680c01e7c9e

**Comment by TheBlueMatt** [comment]
Weird, is it a 0-input transaction (https://github.com/rust-bitcoin/rust-bitcoin/issues/104) cause then the BIP would need updating? Generally I've been using the hex_bytes from bitcoin::util::misc in tests, but I probably should have been using from_hex.

**Comment by TheBlueMatt** [comment]
Ohoh, sorry, I misread. So your problem is you're trying to deserialize the script *with* a length byte, but the script in the BIP does not have its length byte included. You want to just convert the hex string to a Vec<u8> and do Script::from().

**Comment by TheBlueMatt** [comment]
Also note current patch uses spaces instead of tabs in the tests mod. We really need editor configs in these files....

**Comment by savil** [comment]
thank you. Didn't work for me, though. The code I have is:

```
// this line gives error: 
// panicked at 'called `Result::unwrap()` on an `Err` value: OddLength', 
// for hex_str equal to 
// "41044a656f065871a353f216ca26cef8dde2f03e8c16202d2e8ad769f02032cb86a5eb5e56842e92e19141d60a01928f8dd2c875a390f67c1f6c94cfc617c0ea45afaca""
let hex_script: Vec<u8> = decode(hex_str).unwrap();  

let script: Script = Script::from(hex_script);

```

(sorry, I incorrectly pointed out the errant line in my previous gist up top)

**Comment by yuntai** [comment]
@savil 
you seem to have extra 'a' at the end of your `hex_str`.

**Comment by savil** [comment]
[]()

@yuntai much obliged, thanks!

File: src/util/transaction_utils.rs [hunk 1]
```diff
@@ -1,19 +1,92 @@
...
-		} else if b.0.script_pubkey[..] < a.0.script_pubkey[..] { //TODO: ordering of scripts shouldn't be len-based
+		} else if b.0.script_pubkey[..] < a.0.script_pubkey[..] {
 			Ordering::Greater
 		} else {
 			Ordering::Equal
 		}
 	});
 }
+
+// TODO savil. Add tests.
```
Comment (TheBlueMatt) [inline src/util/transaction_utils.rs]: Cool, so still need the tests added here.

File: Cargo.toml [hunk 2]
```diff
@@ -20,6 +20,7 @@ bitcoin = "0.13"
 rust-crypto = "0.2"
 rand = "0.4"
 secp256k1 = "0.9"
+hex = "0.3.2"
```
Comment (savil) [inline Cargo.toml]: I'll make this test-only

**Comment by TheBlueMatt** [comment]
I'm not actually sure if we ever call sort_outputs with different-length scripts, but if we do sort_outputs is clearly wrong - Rust's comparison will sort *first* by length, then by contents, whereas the BIP *seems to* (but doesn't actually explicitly say) that you should sort *first* by contents, *then* by length, as if by memcmp. Hence my comment that the BIP is underspecifed and needs to be fleshed out, with more test vectors added to it.

**Comment by TheBlueMatt** [comment]
But, given the BIP is generally a terrible idea for anything but Lightning, I'd really strongly prefer we update the BOLT to be specified, and not reference the BIP at all.

File: src/util/transaction_utils.rs [hunk 3]
```diff
@@ -1,19 +1,212 @@
...
+	fn sort_output_by_script_pubkey() {
+		let txout1 = TxOut {
+			value:  100,
+			script_pubkey: Builder::new().push_int(3).into_script(),
+		};
+		let txout1_ = txout1.clone();
+
+		let txout2 = TxOut {
+			value: 100,
+			script_pubkey: Builder::new().push_int(1).push_int(2).into_script()
```
Comment (savil) [inline src/util/transaction_utils.rs]: > I'm not actually sure if we ever call sort_outputs with different-length scripts, 

@TheBlueMatt I'm confused. Does this test-case not exercise different-length scripts? if not, could you advise me on how to construct a different-length script?

Comment (TheBlueMatt) [reply src/util/transaction_utils.rs]: Oh, indeed, sorry I missed that, somehow I concluded Rust's comparison did something different.

File: Cargo.toml [hunk 4]
```diff
@@ -27,3 +27,6 @@ gcc = "0.3"
 [dev-dependencies.bitcoin]
 version = "0.13"
 features = ["bitcoinconsensus"]
+
+[dev-dependencies]
+hex = "0.3.2"
```
Comment (TheBlueMatt) [inline Cargo.toml]: I'm generally fine with this, but now we've got two ways of doing hex - using rust-bitcoin's utils one, and this. It probably makes sense to convert them all to the hex crate, cause using an exported rust-bitcoin util method always felt strange to me.

Comment (TheBlueMatt) [reply Cargo.toml]: Also, not gonna block on this, but would be nice to fix in a followup PR.

File: src/util/transaction_utils.rs [hunk 5]
```diff
@@ -1,19 +1,212 @@
...
+	});
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	use bitcoin::blockdata::script::{Script, Builder};
+	use bitcoin::blockdata::transaction::TxOut;
+    use bitcoin::util::hash::Sha256dHash;
```
Comment (TheBlueMatt) [inline src/util/transaction_utils.rs]: There's a few tab/space things in the diff here - I always set my editor to auto-detect whether it should use tabs or spaces based on the file, and also highlight tabs and spaces differently so they're instantly visible.

File: Cargo.toml (hunk 4 above)
Comment (savil) [reply Cargo.toml]: ack, I'll look into it after I wrap up https://github.com/rust-bitcoin/rust-bitcoin/pull/107

File: src/util/transaction_utils.rs (hunk 5 above)
Comment (savil) [reply src/util/transaction_utils.rs]: crap. thought I'd fixed all of these. My apologies.

Comment (TheBlueMatt) [reply src/util/transaction_utils.rs]: No big deal, I'm generally happy to clean up simple stuff like that post-merge (see https://github.com/rust-bitcoin/rust-lightning/pull/90/commits/7f4602596cf9c8d850f514d7b7ca9bc68e95e2dd)

**Pull request merged**